Qemu : Enable SmartFS

manoharahk opened the pull request.

seinfra commented:
Code Rule Check OK.

pradeepns85 commented on os/arch/arm/src/tiva/tiva_flash.c:
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
+	/* Qemu doesnt implement tive flash memory controller.
```
tive -> tiva

manoharahk replied:
Okay, i will correct it

pradeepns85 commented on os/arch/arm/src/tiva/tiva_flash.c:
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
 
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
```
check space after defined

manoharahk replied:
ok

pradeepns85 commented on os/arch/arm/src/tiva/tiva_flash.c:
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
+	/* Qemu doesnt implement tive flash memory controller.
+	 * Hence copy the contents from buf to flash blocks using memcpy.
+	 */
+	memcpy((uint32_t *)(TIVA_VIRTUAL_BASE + startblock * TIVA_FLASH_PAGESIZE), (uint32_t *)buf, nblocks * TIVA_FLASH_PAGESIZE);
```
You could have used src and dest variables declared above . Is there any reason for binding them under not defined flag and not using it inside memcpy ?  May be am wrong. just curious to know

pradeepns85 commented on os/arch/arm/src/tiva/tiva_flash.c:
```diff
@@ -144,11 +144,21 @@ static struct tiva_dev_s g_lmdev = {
 
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
```
check space after defined ..

manoharahk replied:
I will check

pradeepns85 commented on os/arch/arm/src/tiva/chip/lm3s_flash.h:
```diff
@@ -70,11 +70,19 @@
+/* These parts all support a 1KiB erase page size and a total FLASH memory size
+ * of 128MB or 128K pages.
+ */
+#define TIVA_FLASH_NPAGES        (128*1024)
```
Isn't possible to make it dynamic. for ex. if FLASH_SIZE = 128 * 1024 * 1024 ,   then 
#define TIVA_FLASH_NPAGES  FLASH_SIZE/1024    . This way we don't have to modify this for different memory configurations. what's your opinion ?

manoharahk replied:
In my opionion, It just adds one more macro for total flash size then update the existing npage macro. If flash size changes, we need modify flash size macro instead of npage macro.

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+#ifdef CONFIG_MTD_PARTITION
+	mtd_part = mtd_partition(mtd, partoffset, partsize / geo.blocksize, partno);
+	if (!mtd_part) {
+		lldbg("ERROR: failed to create partition.\n");
```
may be you can print like
failed to create  mtd partition and if possible, print the reason like error value?

manoharahk replied:
error value is not returned here

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+
+	if (!mtd) {
+		lldbg("ERROR : up_flashinitializ failed\n");
+		return 0;
```
returning 0 may be for success case , how about returning -1 for failure cases ?

manoharahk replied:
Ok. I will change it to return ERROR.
But, return value is not handled at the caller end.

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+
+	if (mtd->ioctl(mtd, MTDIOC_GEOMETRY, (unsigned long)&geo) < 0) {
+		lldbg("ERROR: mtd->ioctl failed\n");
+		return 0;
```
same as above

manoharahk replied:
ok

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+#endif /* CONFIG_MTD_PARTITION */
+
+	snprintf(partref, sizeof(partref), "p%d", partno);
+	snprintf(devname, sizeof(devname), "/dev/smart%d%s", minorno, partref);
```
Don't we need to check the return value of snprintf ?

manoharahk replied:
It is good to check

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+	ret = mksmartfs(devname, false);
+
+	if (ret != OK) {
+		lldbg("ERROR: mksmartfs on %s failed\n", devname);
```
how about printing the error no?

manoharahk replied:
yes, we can print ret value too.

pradeepns85 commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+	} else {
+		ret = mount(devname, QEMU_SMARTFS_MOUNT_POINT, "smartfs", 0, NULL);
+		if (ret != OK) {
+			lldbg("ERROR: mounting '%s' failed\n", devname);
```
same as above

manoharahk replied:
ok

pillip8282 commented:
@manoharahk 
Could you let me know how to apply patch?
Give me a guide

pillip8282 commented:
@manoharahk 
it is not working... it seems that intiailizing smartfs fail.
belows are log when qemu starts
----------------------------------------------------------------
LM3S6965EVB : FLASH (B) = 134217728 : SRAM (B) = 1048576 : SDRAM (B) = 536870912
ABCDF
smart_findfreephyssector: Program bug!  Expected a free sector 1
board_initialize: ERROR: mksmartfs on /dev/smart0p0 failed
System Information:
	Version: 1.1
	Commit Hash: f2a3514a65a6334c8e9468dcaa05a05fcd6caddf
	Build User: kang@kang-linux
	Build Time: 2018-02-01 08:26:25
	System Time: 08 Jun 2010, 00:00:02 [s] UTC 
TASH>>
----------------------------------------------------------------

Taejun-Kwon commented on os/arch/arm/src/lm3s6965-ek/src/up_boot.c:
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
+
+#define 	QEMU_SMARTFS_PARTITION_START	(0)
+#define 	QEMU_SMARTFS_PARTITION_SIZE	(1024 * 1024) /* 1MB */
```
please check here, tiva using below as a whole flash size.
 #define TIVA_FLASH_NPAGES        256
 #define TIVA_FLASH_PAGESIZE      1024

Taejun-Kwon replied:
Now sflash use TIVA_FLASH_PAGESIZE * TIVA_FLASH_NPAGES = 256K
You set 1024 as a nblocks when call mtd_partition below.
This is something wrong.
To sum up,
1. Total Flash size 256*1024 = 256K
2. Erase Block size = 1024B
3. Number of Erase Block = 256
4. Page size = 1024B
But you defined number of page as a 1024
Please double check this.

manoharahk replied:
TIVA_FLASH_NPAGES is 128K not 256 when running under QEMU.

Please check the changes in " os/arch/arm/src/tiva/chip/lm3s_flash.h"

#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)                        
/* These parts all support a 1KiB erase page size and a total FLASH memory size  
 * of 128MB or 128K pages.                                                          
 */                                                                              
#define TIVA_FLASH_NPAGES        (128*1024)

Taejun-Kwon replied:
128 also problem, total size is TIVA_FLASH_NPAGES * TIVA_FLASH_PAGESIZE.
See ioctl, 
geo->blocksize & geo->erasesize is TIVA_FLASH_PAGESIZE,(Bytes)
geo->neraseblocks is TIVA_VIRTUAL_NPAGES
It means whole of flash is consist of 1024 * 128 = 128K

but you calling below
mtd_partition(mtd, partoffset, partsize / geo.blocksize, partno);
firstblock = (partoffset : 0)
nblocks = partsize/geo.blocksize : (1024 * 1024) / 1024

How allocate 1024K in 128K flash?

manoharahk replied:
Flash size is 128MB in case of QEMU.
Page size is 1024B, Hence TIVA_FLASH_NPAGES = 128 * 1024. =128K.

#define TIVA_VIRTUAL_NPAGES (TIVA_FLASH_NPAGES  - CONFIG_TIVA_FLASH_STARTPAGE).

Where,
CONFIG_TIVA_FLASH_STARTPAGE = 65536 (64K th page). 

Hence,
TIVA_VIRTUAL_NPAGES = 64K

So from 0 - 64MB of flash is not used . From 64MB offset, 1MB flash is used as a smartfs partition.

Taejun-Kwon replied:
you don't understand what I mean..
If below selected by config value, mount will be failed.
#define TIVA_FLASH_NPAGES        256
I am saying that you had better use MTDIOC_GEOMETRY instead of define value, because NPAGES changes significantly based on config.

manoharahk replied:
This code(mtd partition & mount)  is enabled only if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM).

And #if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM),
then  #define TIVA_FLASH_NPAGES        (128*1024)

So How it will fail?

Taejun-Kwon replied:
ah.. ok then I didn't check this file fully, I didn't check whole code is in those configs.
Then It is acceptable.

manoharahk commented:
Change Log:
1. Added a patch to update qemu 16M patch for supporting read/write flash 
2. Error handling for return case
3. snprintf error handling
4. Fixed space issue,
5. Updated memcpy args during tiva_write

manoharahk commented:
@pillip8282 ,

I have updated the  qemu-2.10.0-rc2_16m_ram_size.patch.
Please apply the qemu-2.10.0-rc2_16m_ram_size.patch on qemu-2.10.0-rc2 , build and check.

Steps to build qemu (As documented in build/configs/qemu/README.md):
wget http://download.qemu-project.org/qemu-2.10.0-rc2.tar.xz                     
tar xvJf qemu-2.10.0-rc2.tar.xz                                                  
cd qemu-2.10.0-rc2                                                                                                  
/* If you want to use 16MB RAM, copy qemu-2.10.0-rc2_16m_ram_size.patch from TizenRT/build/configs/qemu */
patch -p1 < qemu-2.10.0-rc2_16m_ram_size.patch                                   
./configure --target-list=arm-softmmu                                            
make -j 4                                                                        
cd arm-softmmu/                                                                  
sudo ln qemu-system-arm /usr/local/bin/qemu-system-arm

Steps to run TizenRT:
qemu-system-arm -M lm3s6965evb -kernel build/output/bin/tinyara -nographic

Expected Result:
LM3S6965EVB : FLASH (B) = 134217728 : SRAM (B) = 16777216 : SDRAM (B) = 536870912
ABCDF
board_initialize: /dev/smart0p0 is mounted successfully at /mnt 
System Information:
	Version: 1.1
	Commit Hash: a764d1e00b0d2c1b98cc4e83b712f7423a680689
	Build User: manohara.hk@manohara-ubuntu
	Build Time: 2018-02-01 09:44:17
	System Time: 08 Jun 2010, 00:00:00 [s] UTC 
TASH>>Hello, World!!


SmartFS Test:
TASH>>smart_test -l 20 /mnt/file1.txt
Creating file /mnt/file1.txt for write mode
Writing test data.  20 lines to write
19
Done.
TASH>>cat /mnt/file1.txt
This is line 0 at offset 0
This is line 1 at offset 27
This is line 2 at offset 55
This is line 3 at offset 83
This is line 4 at offset 111
This is line 5 at offset 140
This is line 6 at offset 169
This is line 7 at offset 198
This is line 8 at offset 227
This is line 9 at offset 256
This is line 10 at offset 285
This is line 11 at offset 315
This is line 12 at offset 345
This is line 13 at offset 375
This is line 14 at offset 405
This is line 15 at offset 435
This is line 16 at offset 465
This is line 17 at offset 495
This is line 18 at offset 525
This is line 19 at offset 555

TASH>>

seinfra commented:
Code Rule Check OK.

manoharahk commented:
Dear pillip8282 ,

Is it working for you now?

pillip8282 commented:
@ manoharahk
Sorry for the late reply.
I've tested it and it's working.

Taejun-Kwon commented on build/configs/qemu/qemu-2.10.0-rc2_16m_ram_size.patch:
```diff
@@ -1,7 +1,7 @@
-From a9f770243ccb354e99b9ff33ecfecbeca73b1b03 Mon Sep 17 00:00:00 2001
+From 23caf033c718dd932efd225a01d00587b88b081a Mon Sep 17 00:00:00 2001
```
Is this file needed??????

manoharahk replied:
yes. Currently flash is made as readonly(ROM) inside qemu. 
Hence i have updated that patch to support flash as "read/write" inside qemu. 
Without that smartfs wont work
